### DISCUSSION TITLE
Do some stuff on SIGTERM *before* the app shuts down

### DISCUSSION THREAD

**Discussion by Jakobhenningjensen** [fastapi/fastapi#11756] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import signal


app = FastAPI()
shutting_down = False

async def do_stuff_before_shutdown():
    shutting_down = True
    # wait 10 seconds before shutting the app down    
    # i.e allowing requests for 10 more seconds
     asyncio.sleep(10) 

@app.get("/health/ready", response_class=PlainTextResponse)
def health_ready():

    text = "OK"
    if shutting_down:
        return PlainTextResponse(text, 425)
    return PlainTextResponse(text, 200)
```

### Description

My FastAPI is run (uvicorn) in a Kubernetes cluster. 
I'm trying to gracefully shut it down when the cluster is scaling down.

What I want to do is when I get a sigterm

1) Change "health/ready" to return 425 (such that Kubernetes does not send any more requests to this pod)
2) Allow the app to run for max 30 seconds to finish current tasks before killing it

losely speaking - I want to run a function when `SIGTERM` is received without "messing" with the sigterm-handler

Step (1) can be done by creating a function and pass it to `signal.signal(signal.SIGTERM, ...)` but that is being called after the "shutdown" event in the lifespan. I have thought about setting it in the `lifespan` event, but as far as I understand that happens after the app is shut down - I want to do it before the app shuts down .

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.7.4

### Python Version

3.11.8

**Comment by aiguofer** [comment]
We are running into the same issue. For more context, see https://github.com/kubernetes/kubernetes/issues/86280#issuecomment-583173036 and https://learnk8s.io/graceful-shutdown.

It seems this is not doable from the `uvicorn` side either: https://github.com/encode/uvicorn/discussions/2257

**Reply by aiguofer** [reply]
@Jakobhenningjensen You should be able to simply add a `preStop` hook to sleep for a few seconds. This will give k8s enough time to remove the pod from most proxies before actually sending the SIGTERM to your process: https://freecontent.manning.com/handling-client-requests-properly-with-kubernetes/

**Reply by Jakobhenningjensen** [reply]
I'll look into that.
Currently I've (as written below) forked uvicorn  and implemented a "shutdown delay" which we are currently using 

https://github.com/Jakobhenningjensen/uvicorn

But I think I'll look at the `preStop` and see if that can do it!

**Comment by dbfreem** [comment]
I am not quite following why wouldn't catching the k8s `signals.signals(signal.SIGTERM, ..........)` not work? Wouldn't that get called before the app shutdown with lifespan?

**Reply by Jakobhenningjensen** [reply]
I have tried to catch that aswell but then you end up overwriting unicorns signal handler, and then your app won't shutdown at all.
I have tried many(!) workarounds regarding catching the signal but I haven't been able to make it work.

Currently I've forked unicorn and implemented a solution there https://github.com/Jakobhenningjensen/uvicorn and I'm working on submitting a PR (I'm currently struggling with a unit test). 
We are using that exact code in production though and it seems to work great.

**Reply by dbfreem** [reply]
ahhhh I see now catching the SIGTERM causes issues with uvicorn. I wasn't aware of that.

**Reply by LexFuturorum** [reply]
sigterm is not catchable, that is linux behavior,
im not familiar with k8s, but if possible you could try SIGINT as that is catchable

**Comment by XinEDprob** [comment]
For step 1 in the issue description
> Change "health/ready" to return 425 (such that Kubernetes does not send any more requests to this pod)

from the lifespan of asgi doc (https://asgi.readthedocs.io/en/latest/specs/lifespan.html#shutdown-receive-event), the shutdown of liftspan happens after ```server has stopped accepting connections and closed all active connections```. Does this mean that we do not need to change "health/ready" to return 425 (step 1 in the issue description) before the "shutdown" event in the lifespan? fastapi does not accept new requests anyway.
For step 2

> Allow the app to run for max 30 seconds to finish current tasks before killing it

I did some experiments and noticed that after fastapi receives the sig-term and assuming no sig-kill command, only if all the in-flight requests are processed in fastapi, the fastapi starts to shut down. Does this mean step 2 is also perfectly handled by fastapi? 

I am a newbie to fastapi and not sure if above arguments are correct or not. If they are correct, seems there is no need to do these two steps before the shutdown event in the lifespan.

**Comment by bentosming** [comment]
Dne st 30. 4. 2025 23:02 uživatel XinEDprob ***@***.***>
napsal:

**Comment by alv2017** [comment]
@Jakobhenningjensen 
I had a look at your code, and it seems that the functionality you are trying to implement is already there: https://www.uvicorn.org/settings/#timeouts

Or you are trying to achieve something else?

**Reply by Jakobhenningjensen** [reply]
> @Jakobhenningjensen 
> I had a look at your code, and it seems that the functionality you are trying to implement is already there: https://www.uvicorn.org/settings/#timeouts
> 
> Or you are trying to achieve something else?

The graceful-shutdown just allows running request to finish. What I want to do is to delay the shutdown, while responding 425 instead of 200 on health requests for the next 10 seconds. This allows load balancers to register that an instance is under shutdown, and can remove the instance from the pool og available instances.

**Comment by mdaif** [comment]
I have the exact same issue. In Sanic, this case could be handled by registering a listener of the `before_server_stop` signal 
```
app.register_listener(on_before_server_stop, "before_server_stop")
```
and then in the handler:
```
def on_before_server_stop(app: Sanic, loop: asyncio.BaseEventLoop) -> None:
  app.ctx.is_shutting_down = True
```
Which could be checked later in the readiness check endpoint.

My understanding is, this is possible in Sanic because it's both the framework and the server, while FastAPI is the framework.

I hope someone from the team could help us with this question. It's really crucial, especially for scaling down events.

**Reply by Jakobhenningjensen** [reply]
Yeah, exactly. I figure out that the "issue" lies in uvicorn, used to run Fastapi. I managed to implement a solution which works as a forked version of FastAPI

**Reply by mdaif** [reply]
AWS recommendation for this situation (the framework doesn't gracefully handle the SIGTERM), is to use the preStop hook.
https://docs.aws.amazon.com/eks/latest/best-practices/load-balancing.html#_gracefully_handle_the_client_requests

**Comment by kkpattern** [comment]
We have a similar issue. Our application relies heavily on SSE. Since these are long-lived connections, they prevent the server from draining naturally. Without a 'before shutdown' hook to proactively close them, the graceful shutdown process hangs and almost inevitably leads to a timeout.
